File: pypers/api/services.py
""" 
 This file is part of Pypers.

 Pypers is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 Pypers is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Pypers.  If not, see <http://www.gnu.org/licenses/>.
 """

from flask import request
from flask.ext.restful import reqparse, Resource
from pypers.utils.utils import subproc
from pypers import pipelines

from supervisor.options import ClientOptions
from supervisor.supervisorctl import Controller


def get_supervisor_status(name):
    """
    Get the status of a supervisor service
    """
    res = subproc("supervisorctl status %s" % name)
    status =  True if res['stdout'].lower().find("running") > -1 else False
    return status


class ServicesApi:
    SERVICES = ['demux']

    class ServicesList(Resource):
        """
        This class return the list of pipelines types
        """
        def get(self):
            service_status = {}
            for s in ServicesApi.SERVICES:
                service_status[s] = get_supervisor_status(s)

            print service_status
            return service_status


    class ServiceStatus(Resource):
        """
        This class is the interface to execute  pipeline
        """

        def get(self, name):
            """
            Return the list of parameter for a specifc pipeline
            """
            return {"status": get_supervisor_status(name)}

        def put(self, name):
            """
            Return the list of parameter for a specifc pipeline
            """


            data = request.get_json()
            status = bool(data.get('status'))

            cmd = "start" if status else "stop"
            res = subproc("supervisorctl %s %s" % (cmd, name))

            print res['stdout']
            ret_code = 200 if res['stdout'].lower().find(cmd) > -1 else 500
            return {"msg" : res['stdout']}, ret_code

    @classmethod
    def connect(cls, api):
        """
        Connect the end points to corresponding classmethod
        """

        api.add_resource(cls.ServicesList,  '/api/services')
        api.add_resource(cls.ServiceStatus, '/api/services/<string:name>')

File: mithril-ui/src/scripts/pages/services.js
 // This file is part of Pypers.

 // Pypers is free software: you can redistribute it and/or modify
 // it under the terms of the GNU General Public License as published by
 // the Free Software Foundation, either version 3 of the License, or
 // (at your option) any later version.

 // Pypers is distributed in the hope that it will be useful,
 // but WITHOUT ANY WARRANTY; without even the implied warranty of
 // MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 // GNU General Public License for more details.

 // You should have received a copy of the GNU General Public License
 // along with Pypers.  If not, see <http://www.gnu.org/licenses/>.

var m = require('mithril');
    _ = require('lodash');

var utils   = require('../utils');

var NesLayout   = require('../components/neslayout')

var ServiceStatus = {
    list: function() {
        return m.request({
            method: 'GET',
            url: '/api/pypers/services',
            extract: utils.request.authenticate
        });
    },
    set: function(name, status){
        return m.request({
            method: 'PUT',
            url: '/api/pypers/services/' + name,
            extract: utils.request.authenticate,
            data: { 'status' : status }
        });
    },
}


ServicePage = {
    controller: function() {
        var ctrl = this;
        ctrl.serviceStatusList = ServiceStatus.list()
        ctrl.toggleStatus = function(name, status){
            ServiceStatus.set(name, ! status)
            ctrl.serviceStatusList = ServiceStatus.list()
        }

    },
    view: function(ctrl) {
        return (
            m.component(NesLayout, {
                menu: 'Services',
                breadcrumbs: [{label: 'Services'}],
                main: [
                    m('ul',
                        _.map(ctrl.serviceStatusList(), function(value, key){
                            return (
                                m('li', {class:'cgi-services'},
                                    m('.cgi-services-label', key),
                                        m('.cgi-services-status switch',
                                        m('input', {
                                            id: 'demultiplexing_toggle_switch', 
                                            checked: value, 
                                            class:'cmn-toggle cmn-toggle-round-flat', 
                                            type:'checkbox',
                                            onclick : ctrl.toggleStatus.bind(ctrl, key, value)}
                                        ),
                                        m('label', {for: 'demultiplexing_toggle_switch'})
                                    )
                                )
                            )
                        })
                    )
                ]
            })
        );
    }
};

exports.view       = ServicePage.view
exports.controller = ServicePage.controller

